x/concentrated-liquidity/swaps.go

```diff
@@ -134,6 +182,30 @@ func (k Keeper) SwapOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coin, tokenOutDen
 	return tokenOut, nil
 }
 
+func (k *Keeper) SwapInAmtGivenOut(
+	ctx sdk.Context,
+	desiredTokenOut sdk.Coin,
+	tokenInDenom string,
+	swapFee sdk.Dec,
+	priceLimit sdk.Dec,
+	poolId uint64,
+) (tokenIn sdk.Coin, err error) {
+	tokenIn, actualTokenOut, updatedTick, updatedLiquidity, updatedSqrtPrice, err := k.calcInAmtGivenOut(ctx, desiredTokenOut, tokenInDenom, swapFee, priceLimit, poolId)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	if err := k.applySwap(ctx, tokenIn, actualTokenOut, poolId, updatedLiquidity, updatedTick, updatedSqrtPrice); err != nil {
```

Comment:
We should not have this here.

The caller of this function - `SwapExactAmountOut` checks against `tokenInMaxAmount` and fails if the max is exceeded.

However, here we might be mutating the state before checking against max.

In general, there is no reason for having 2 swap methods for each kind of swap. We should do the following - remove `SwapInAmtGivenOut` and `SwapOutAmtGivenIn`. Move all logic to `SwapExactAmountOut` and `SwapExactAmountIn` respectively.

After moving, call `applySwap` after we check against the `tokenInMaxAmount` and `tokenOutMinAmount` respectively.

Reply:
In this PR, let's move the call to `k.applySwap()` to `k.updatePoolForSwap()`.

Let's create an issue for deleting swap methods in a separate PR please

Reply:
This is following the design of what we do for Balancer. Additionally, we updatePoolForSwap in the SwapExact methods. Shouldn't we keep the design as similar as possible between the pool types to prevent confusion?

Reply:
It's incorrect to update and persist the pool state when we end up not doing the swap due to price impact protection. That should be changed.

We should aim to follow balancer implementation only when it makes sense. There is no reason for having 2 swap methods. The balancer/gamm has 2 because there is one at the keeper layer and one at the pool layer. Here, we have 2 on the keeper layer when one would suffice.

Reply:
I didnt see this comment, let me address this before last review